geometry/meshop.cpp

```diff
@@ -390,6 +502,162 @@ Mesh loadObj(const boost::filesystem::path &filename, ObjMaterial *mtl)
+                         math::Points3i* faceColors,
+                         math::Points3* vertexNormals,
+                         std::vector<int>* faceLabels)
+{
+    std::ifstream f(filename.native());
+    if (!f.good())
+    {
+        LOGTHROW(err2, std::runtime_error) << "Can't open " << filename;
+    }
+
+    f.exceptions(std::ios::badbit | std::ios::failbit);
+
+    // read header
+    std::string line;
+
+    bool hasVertexColors = false;
+    bool hasAlphaVertexColor = false;
+    bool hasVertexNormals = false;
+    bool hasFaceColors = false;
+    bool hasAlphaFaceColor = false;
+    bool hasFaceLabels = false;
+    int nvert = -1;
+    int ntris = -1;
+
+    do
+    {
+        if (getline(f, line).eof()) { break; }
+        boost::algorithm::trim(line);
+        if (sscanf(line.c_str(), "element vertex %d", &nvert))
+        {
+            if (getline(f, line).eof()) { break; }
+            boost::algorithm::trim(line);
+            utility::expect(hasProperties(f, line, { "x", "y", "z" }),
+                            "Vertices have to have properties x, y, z");
+            if (hasProperties(f, line, { "red", "green", "blue" }))
+            {
+                hasVertexColors = true;
+            }
+            if (hasProperties(f, line, {"alpha"}))
+            {
+                hasAlphaVertexColor = true;
+            }
+            if (hasProperties(f, line, { "nx", "ny", "nz" }))
+            {
+                hasVertexNormals = true;
+            }
+        }
+
+        if (sscanf(line.c_str(), "element face %d", &ntris))
+        {
+            if (getline(f, line).eof()) { break; }
+            boost::algorithm::trim(line);
+            utility::expect(line == "property list uchar int vertex_indices",
+                            "Faces have to have list property");
+            if (getline(f, line).eof()) { break; }
+            boost::algorithm::trim(line);
+            if (hasProperties(f, line, { "red", "green", "blue" }))
+            {
+                hasFaceColors = true;
+            }
+            if (hasProperties(f, line, {"alpha"}))
+            {
+                hasAlphaFaceColor = true;
+            }
+            if (hasProperties(f, line, { "label" })) { hasFaceLabels = true; }
+        }
+    }
+    while (line != "end_header");
+
+    if (nvert < 0 || ntris < 0)
+    {
+        LOGTHROW(err2, std::runtime_error)
+            << filename << ": unknown PLY format.";
+    }
+
+    Mesh mesh;
+    mesh.vertices.reserve(nvert);
+
+    // load points
+    for (int i = 0; i < nvert; i++)
+    {
+        double x, y, z;
+        f >> x >> y >> z;
+        if (hasVertexColors)
+        {
+            int r, g, b;
+            f >> r >> g >> b;
+            if (vertexColors) { vertexColors->emplace_back(r, g, b); }
+        }
+        if (hasAlphaVertexColor)
+        {
+            int al;
+            f >> al;
+        }
+        if (hasVertexNormals)
```

Comment:
The PLY header should define the structure of vertex/face. With this implementation, we are only "allowing" PLY files that are formatted in a very specific way. I would say that this function should be more robust and not restrict the order of the properties, rather use the information from the header to parse it.

Reply:
My intention was not to write a full-blown ply parser. I mention this restriction in the docstring.
If you think it's necessary, I can rewrite it to support arbitrary property permutation. But still, it won't be able to support an arbitrary list of properties.

Reply:
@vaclavblazek What do you think about it?

Reply:
I see two possibilities - either move this function to an internal library or find some library for proper ply parsing and use that.

Reply:
Parser should be as generic as viable. We do not need to support any exotic stuff but should not depend on specially crafted format.

Reply:
As discussed with Vašek, I will rewrite the parser as a generic `ply` reader that calls user-defined callbacks for each vertex/face.

Reply:
It took a while, but I have the general ply parser implemented in `parse-ply.hpp` and `parse-ply.cpp`.
https://github.com/melowntech/libgeometry/blob/tomasn/ply-save-features/geometry/parse-ply.cpp
However, it has one drawback - parsing a ply file without any face/vertex features takes about twice as long as `loadPly()`.  The initial plan was to replace `loadPly()` by some function using this parser - experimental version is here:
https://github.com/melowntech/libgeometry/blob/2c013bf9404054163d88d63f8d15d8f70520b877/geometry/parse-ply.cpp#L251

I *guess* the slowdown is caused by copying lines to stringsteam - valgrind indicates twice as many memory allocations. The copy is done here:
https://github.com/melowntech/libgeometry/blob/2c013bf9404054163d88d63f8d15d8f70520b877/geometry/parse-ply.cpp#L310
Note that it cannot be easily replaced by move, because std::istringstream constructor does not support it.

I am not sure what to do with this situation. Either I can leave `loadPly()` as is and keep the separate `parse-ply.hpp` module or I can replace `loadPly()` and cause slowdown in various places.

Reply:
Why not parse data directly from the input file stream? What are the advantages of using getline, then constructing stringstream and parsing from it? Note: another problem with the new implementation is that the state of the stringstream is not checked whatsoever which may lead to uninitialized values if the input file is corrupted.

Reply:
If it's really needed; maybe you could use `boost::iostreams::array_source` which can operate on an external (not owned) buffer. @tomas2211

Reply:
Good suggestion, I changed it. Parsing now takes about the same time as the old `loadPly()` so I changed it to use the new parser.